Change default file format to 1.8

fortnern opened the pull request:
2 remaining test failures seemingly related to cmake. Adding this PR in the hopes that someone better at cmake than me can help.

Closes #5315 
<!-- ELLIPSIS_HIDDEN -->

----

> [!IMPORTANT]
> Change default file format to 1.8 across various tests and examples, updating file creation and access logic accordingly.
> 
>   - **Behavior**:
>     - Default file format version changed to 1.8 in `H5Pfapl.c`.
>     - Updated file creation and access to use 1.8 format in `h5ex_g_compact.c` and `test_file_image.c`.
>     - Set earliest file format in multiple test files including `cache_tagging.c`, `dtypes.c`, and `links.c`.
>   - **Tests**:
>     - Modified expected output in `tools/test/misc/expected/*.ls` files to reflect new file format locations.
>     - Adjusted test logic in `test_file_image.c` and `cache_tagging.c` to accommodate format changes.
>   - **Misc**:
>     - Added comments and TODOs for future format testing in `test_file_image.c`.
>     - Minor variable renaming for clarity in `test_file_image.c`.
> 
> <sup>This description was created by </sup>[<img alt="Ellipsis" src="https://img.shields.io/badge/Ellipsis-blue?color=175173">](https://www.ellipsis.dev?ref=HDFGroup%2Fhdf5&utm_source=github&utm_medium=referral)<sup> for 85332eddfbd14d8d1f47b3cbe75d28b58fecda33. You can [customize](https://app.ellipsis.dev/HDFGroup/settings/summaries) this summary. It will automatically update as commits are pushed.</sup>

<!-- ELLIPSIS_HIDDEN -->

fortnern committed:
Update default file format verison to 1.8. Not all tests are passing
yet.

github-actions[bot] committed:
Committing clang-format changes

fortnern committed:
Merge branch 'develop' into def_ff_18

fortnern committed:
Fix remaining errors except for 2 h5import failures.

github-actions[bot] committed:
Committing clang-format changes

fortnern committed:
typo

Co-authored-by: ellipsis-dev[bot] <65095814+ellipsis-dev[bot]@users.noreply.github.com>

fortnern committed:
Fix more test failures

fortnern committed:
Merge branch 'def_ff_18' of github.com:fortnern/hdf5 into def_ff_18

fortnern committed:
Fix direct I/O test failure

byrnHDF commented:
There were no CMake changes so need to see the failing instance.

byrnHDF commented:
One of the first workflows failing had two test failures:
The following tests FAILED:
	640 - H5IMPORT_SUB-DUMP-tall_fp32-CMP (Failed)
	645 - H5IMPORT_SUB-DUMP-tall_i32-CMP (Failed)

byrnHDF commented:
Current CI [CMake 4 arm64 Workflows / MacOS Clang-Release--](https://github.com/HDFGroup/hdf5/actions/runs/18856498895/job/53805634414#logs) has these tests failing:

The following tests FAILED:
	2707 - JUnit-TestH5Fparams (Failed)
	2708 - JUnit-TestH5Fbasic (Failed)
	2727 - JUnit-TestH5P (Failed)
	2729 - JUnit-TestH5Pfapl (Failed)
Errors while running CTest

byrnHDF commented:
Searching the logs for the failing tests shows this for Fparams java test:
JUnit version 4.13.2
2599: .testH5Fis_accessible_null
2599: .testH5Fcreate_null
2599: .testH5Fflush_local
2599: .testH5Fget_info
2599: E.testH5Fmount_null
2599: .testH5Fcreate
2599: .testH5Fflush_global
2599: .testH5Funmount_null
2599: .testH5Fclose_negative
2599: .testH5Fopen_null
2599: 
2599: Time: 0.064
2599: There was 1 failure:
2599: 1) testH5Fget_info(test.TestH5Fparams)
2599: java.lang.AssertionError: H5.H5Fget_info: java.lang.AssertionError: expected:<2> but was:<0>
2599: 	at org.junit.Assert.fail(Assert.java:89)
2599: 	at test.TestH5Fparams.testH5Fget_info(TestH5Fparams.java:218)
2599: 
2599: FAILURES!!!
2599: Tests run: 10,  Failures: 1

fortnern commented:
> There were no CMake changes so need to see the failing instance.



> One of the first workflows failing had two test failures: The following tests FAILED: 640 - H5IMPORT_SUB-DUMP-tall_fp32-CMP (Failed) 645 - H5IMPORT_SUB-DUMP-tall_i32-CMP (Failed)

This was the failure that I thought was cmake related - those two test files were missing. It's possible that my working directory was messed up. I deleted it and tried again and saw the same two tests fail, but it's possible I didn't notice they failed for a different reason, as Jordan saw the same two tests fail with an expected output mismatch. I fixed the mismatch so they should work now. Do they still fail?

fortnern committed:
Fix Java tests I hope

fortnern committed:
Changelog.md note

byrnHDF reviewed (dismissed):
Java changes looks good and CI is passing

mattjala commented on test/vfd.c:
Nitpick, but this comment should be updated

fortnern commented on test/vfd.c:
```diff
@@ -772,7 +772,7 @@ test_direct(void)
      * 4 currently and the size of the file should be between 3 & 4 file buffer
```
```suggestion
     * 4 currently and the size of the file should be between 1 & 2 file buffer
```

fortnern committed:
Update comment in test/vfd.c

jhendersonHDF commented on hl/test/test_file_image.c:
```diff
@@ -99,6 +103,13 @@ test_file_image(size_t open_images, size_t nflags, const unsigned *flags)
         FAIL_PUTS_ERROR("malloc() failed");
 
+    /* Create FAPL and set earliest format */
+    /* TODO: run this test with all different formats */
```
TODO still relevant here?

fortnern replied:
Yes. It will require different hacking of the superblock for each format version which is why I didn't have time to do it for this release. I can file an issue for it.

jhendersonHDF commented on test/tfile.c:
```diff
@@ -8262,6 +8283,12 @@ test_deprec(const char *driver_name)
+        /* Close temp FAPLs */
+        ret = H5Pclose(tmp_fapl);
+        CHECK(ret, FAIL, "H5Pclose");
+        ret = H5Pclose(new_fapl);
```
It looks like `new_fapl` is never created or copied from another FAPL above. Should it and this close call just be removed?

fortnern replied:
`new_fapl` is retrieved on line 8219 and queried on line 8220. `new_fapl` was present before, this change only changes the scope of the variable and adds the missing close. It's ok to have the restricted scope because, since this is in testhdf5, it doesn't bail out if it hits a failure.

jhendersonHDF replied:
Ah, missed the line above removing `new_fapl` from its current spot

brtnfld reviewed (dismissed):
Only checked Fortran

hyoklee commented:
@fortnern, FYI - I see 2 VFD test failures with HL & CPP & all debugging & v16 API options on Ubuntu-24.04:

https://my.cdash.org/viewTest.php?onlyfailed&buildid=3167967

I hope it's not related to your change.

jhendersonHDF commented:
Just needs the breaking change section added to the changelog then looks good

fortnern commented:
> @fortnern, FYI - I see 2 VFD test failures with HL & CPP & all debugging & v16 API options on Ubuntu-24.04:
> 
> https://my.cdash.org/viewTest.php?onlyfailed&buildid=3167967
> 
> I hope it's not related to your change.

I configured HDF5 with HL, CPP, and 1.6 API, and cannot reproduce these failures with:
`HDF5_TEST_DRIVER="multi" ./cpp_testhdf5`
or
`HDF5_TEST_DRIVER="family" ./cpp_testhdf5`

What do you mean by "all debugging"?

Is this CDash test testing my branch? This hasn't been merged to develop yet.

fortnern committed:
Add breaking change note

hyoklee commented:
> > @fortnern, FYI - I see 2 VFD test failures with HL & CPP & all debugging & v16 API options on Ubuntu-24.04:
> > https://my.cdash.org/viewTest.php?onlyfailed&buildid=3167967
> > I hope it's not related to your change.
> 
> I configured HDF5 with HL, CPP, and 1.6 API, and cannot reproduce these failures with: `HDF5_TEST_DRIVER="multi" ./cpp_testhdf5` or `HDF5_TEST_DRIVER="family" ./cpp_testhdf5`
> 
> What do you mean by "all debugging"?
> 
> Is this CDash test testing my branch? This hasn't been merged to develop yet.

Yes. You can easily duplicate it using my GH Action script:
[lin.yml](https://raw.githubusercontent.com/hyoklee/actions/ad34b8ccb840224be3f8cd94750ed0697efb2018/.github/workflows/lin.yml)

jhendersonHDF commented:
> cannot reproduce these failures with: `HDF5_TEST_DRIVER="multi" ./cpp_testhdf5` or `HDF5_TEST_DRIVER="family" ./cpp_testhdf5`

You may need to use `HDF5_DRIVER="multi" ./cpp_testhdf5` / `HDF5_DRIVER=family" ./cpp_testhdf5` instead. `HDF5_TEST_DRIVER` is mostly meant for tests that use `h5_fileaccess()` to get a FAPL (sort of a leftover from Autotools) where `HDF5_DRIVER` applies to all FAPLs by default.

fortnern commented:
These failures appear to be preexisting bugs that were exposed by the change, and not simply tests expecting the oldest file format. I'm trying to track them down but it might take some time.

fortnern committed:
Fix problems with family driver with user block. Fix bugs in testhdf5.cpp.

fortnern commented:
The 2 failures with multi were due to issues with the test code. The family failure was due to weirdness in how the family driver handled the user block. My position is the family driver shouldn't handle it at all, but I'm willing to hear other opinions. Either way the way it handled it before was wrong.

jhendersonHDF commented on release_docs/CHANGELOG.md:
```diff
@@ -59,6 +63,10 @@ For releases prior to version 2.0.0, please see the release.txt file and for mor
 
+### Fixed problems with family driver and user block
+
+   When using a user block with the family driver, the driver would inappropriately subtract the user block size for each member file when calculating member eoas. This could cause a failure when an addresse overflowed the calculated eoa. The driver would also add the user block size when returning the eof. Modified the family driver to not consider the user block, as it is handled by the H5FD la ...
```
```suggestion
   When using a user block with the family driver, the driver would inappropriately subtract the user block size for each member file when calculating member eoas. This could cause a failure when an address overflowed the calculated eoa. The driver would also add the user block size when returning the eof. Modified the family driver to not consider the user block, as it is handled by the H5FD layer. The user block now spans the first x bytes of the family array, for example a 4 KiB user block with 3 KiB member size will take up the entire first member and the first 1 KiB of the second. This may cause compatibility issues with preexisting family files with user blocks, though the way it worked before was inconsistent if it worked at all.
```

jhendersonHDF commented on release_docs/CHANGELOG.md:
```diff
@@ -557,6 +569,10 @@ Added Fortran wrapper h5fdsubfiling_get_file_mapping_f() for the subfiling file
 
+### Fixed problems with family driver and user block
+
+   When using a user block with the family driver, the driver would inappropriately subtract the user block size for each member file when calculating member eoas. This could cause a failure when an addresse overflowed the calculated eoa. The driver would also add the user block size when returning the eof. Modified the family driver to not consider the user block, as it is handled by the H5FD la ...
```
```suggestion
   When using a user block with the family driver, the driver would inappropriately subtract the user block size for each member file when calculating member eoas. This could cause a failure when an address overflowed the calculated eoa. The driver would also add the user block size when returning the eof. Modified the family driver to not consider the user block, as it is handled by the H5FD layer. The user block now spans the first x bytes of the family array, for example a 4 KiB user block with 3 KiB member size will take up the entire first member and the first 1 KiB of the second. This may cause compatibility issues with preexisting family files with user blocks, though the way it worked before was inconsistent if it worked at all.
```

jhendersonHDF commented on c++/test/tfile.cpp:
```diff
@@ -80,7 +80,7 @@ test_file_create()
     try {
         // Create file FILE1
-        file1 = new H5File(FILE1, H5F_ACC_EXCL);
+        file1 = new H5File(FILE1, H5F_ACC_TRUNC);
```
Why the change in file access mode here?

fortnern replied:
The test is creating a new file. H5F_ACC_EXCL causes it to fail if the file already exists, which is what was happening. It was failing silently before, but with the 1.8 format it causes issues due to a fapl mismatch - see #5954

jhendersonHDF replied:
I see. Fine either way, though previously the issues with this test involved the fact that it left a file open unintentionally and also forgot to delete it, so I wonder if that isn't the case here as well.

fortnern replied:
I think this is at least partly due to it simply using HDremove() to remove the test files, which obviously doesn't work with the family file driver. Something that should be fixed for sure, but not within the scope of this PR

lrknox committed:
Merge branch 'develop' into def_ff_18

mattjala reviewed (changes requested):
Address typos

fortnern committed:
Update changelog.md notes and fix spelling

fortnern committed:
Merge branch 'def_ff_18' of github.com:fortnern/hdf5 into def_ff_18

fortnern committed:
More typos

fortnern commented:
We talked with Quincey about the family driver changes and he agreed with this approach.

lrknox commented:
Matt's review was dismissed by the commit of a typo fix in CHANGELOG.md
